DNN Platform/Library/ExtensionPoints/SafeDirectoryCatalog.cs

```diff
@@ -34,19 +35,51 @@ public class SafeDirectoryCatalog : ComposablePartCatalog
 
         public SafeDirectoryCatalog(string directory)
         {
-            var files = Directory.EnumerateFiles(directory, "*.dll", SearchOption.AllDirectories);
-
             _catalog = new AggregateCatalog();
 
+            // ASP.NET will load all the DLL files under the bin folder; the first loop would be sufficient
+            var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies()
+                                            .Where(asm => asm.GetName().CodeBase.StartsWith("file://"));
+
+            var processedFiles = new List<string>();
+            foreach (var assembly in loadedAssemblies)
+            {
+                var fileUri = new Uri(assembly.GetName().CodeBase);
+                var fi = new FileInfo(fileUri.AbsolutePath);
+                if (fi.DirectoryName != null &&
+                    fi.DirectoryName.StartsWith(directory, StringComparison.InvariantCultureIgnoreCase))
+                {
+                    try
+                    {
+                        processedFiles.Add(fi.FullName.ToUpper());
+                        var asmCatalog = new AssemblyCatalog(assembly);
+                        //Force MEF to load the plugin and figure out if there are any exports
+                        // good assemblies will not throw the RTLE exception and can be added to the catalog
+                        if (asmCatalog.Parts.ToList().Count > 0) _catalog.Catalogs.Add(asmCatalog);
+                    }
+                    catch (ReflectionTypeLoadException)
+                    {
+                    }
+                    catch (BadImageFormatException)
+                    {
+                    }
+                    catch (FileLoadException) //ignore when the assembly load failed.
+                    {
+                    }
+                }
+            }
+
+            // process whatever left in subfolders
+            var files = Directory.EnumerateFiles(directory, "*.dll", SearchOption.AllDirectories)
+                                 .Select(f => f.ToUpper()).Except(processedFiles);
             foreach (var file in files)
             {
                 try
                 {
-	                var asmCat = new AssemblyCatalog(file);
-
-	                //Force MEF to load the plugin and figure out if there are any exports
-	                // good assemblies will not throw the RTLE exception and can be added to the catalog
-	                if (asmCat.Parts.ToList().Count > 0) _catalog.Catalogs.Add(asmCat);
+                    var asmCatalog = new AssemblyCatalog(file);
+                    //Force MEF to load the plugin and figure out if there are any exports
+                    // good assemblies will not throw the RTLE exception and can be added to the catalog
+                    if (asmCatalog.Parts.ToList().Count > 0) _catalog.Catalogs.Add(asmCatalog);
```

Comment:
I realize this isn't your code, but it seems like the `ToList()` is unnecessary here, I would expect that `asmCatalog.Parts.Any()` would be more efficient